# <font color=blue> 练习2. 用CNN网络对手语数据集的图像分类 <blue>

## <font color=blue>1 目标 </font>

在完成本节时，您将能够：
* 学会使用kaggle在线数据集
* 自己定义一个CNN模型
* 训练CNN模型并观察结果

## <font color=red>实验完成代码扣掉的部分：</font> <br>
#################### <br>
#完善代码<br>
####################<Br>

In [1]:
import torch.nn as nn
import pandas as pd
import torch
from torch.optim import Adam
from torch.utils.data import Dataset, DataLoader



import os 
os.environ['CUDA_LAUNCH_BLOCKING']='1'

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
torch.cuda.is_available()

True

## <font color=blue>2 加载数据集 </font>

In [2]:
#产看数据集目录情况
print(os.listdir('../input'))

['sign-language-mnist-24label']


In [4]:
train_df = 
valid_df = 
####################
#完善代码
####################

提取前5行数据观察

In [5]:
####################
#完善代码
####################

array([[107, 118, 127, ..., 204, 203, 202],
       [155, 157, 156, ..., 103, 135, 149],
       [187, 188, 188, ..., 195, 194, 195],
       [211, 211, 212, ..., 222, 229, 163],
       [164, 167, 170, ..., 163, 164, 179]])

In [6]:
sample_x.shape

(5, 784)

上面 28x28 像素的仍然是按照向量表示，不是图片，丢失像素位置和关联信息

对于我们模型的第一个卷积层，我们不仅需要图像的高度和宽度，还需要颜色通道的数量。我们的图像是灰度的，所以只有 1 个通道。

这意味着需要将当前形状 (5, 784) 转换为 (5, 1, 28, 28)。使用 NumPy 数组，我们可以为任何我们希望保持不变的维度传递 -1。

In [7]:
####################
#完善代码
####################

sample_x = sample_x.reshape(-1, IMG_CHS, IMG_HEIGHT, IMG_WIDTH)
sample_x.shape

(5, 1, 28, 28)

## <font color=blue>3 创建数据读取类 </font>

In [8]:
IMG_HEIGHT = 28
IMG_WIDTH = 28
IMG_CHS = 1
class MyDataset(Dataset):
####################
#完善代码
####################

注意数据加载器中，什么函数调用需要调用 shuffle=True 参数？

In [9]:
BATCH_SIZE = 8

train_data = MyDataset(train_df)
####################
#完善代码
####################
valid_data = MyDataset(valid_df)
valid_loader = DataLoader(valid_data, batch_size=BATCH_SIZE)
valid_N = len(valid_loader.dataset)

## <font color=blue>4 创建CNN模型 </font>

In [10]:
n_classes = 24
kernel_size = 3
flattened_img_size = 75 * 3 * 3

model = nn.Sequential(
    # First convolution
    nn.Conv2d(IMG_CHS, 25, kernel_size, stride=1, padding=1),  # 25 x 28 x 28
    nn.BatchNorm2d(25),
####################
#完善代码
####################
    nn.Dropout(.3),
    nn.ReLU(),
    nn.Linear(512, n_classes)
)

## <font color=blue>5 训练模型 </font>

In [11]:
def get_batch_accuracy(output, y, N):
    pred = output.argmax(dim=1, keepdim=True)
    correct = pred.eq(y.view_as(pred)).sum().item()
    return correct / N

In [12]:
model.to(device)

Sequential(
  (0): Conv2d(1, 25, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
  (1): BatchNorm2d(25, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (2): ReLU()
  (3): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
  (4): Conv2d(25, 50, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
  (5): BatchNorm2d(50, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (6): ReLU()
  (7): Dropout(p=0.2, inplace=False)
  (8): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
  (9): Conv2d(50, 75, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
  (10): BatchNorm2d(75, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (11): ReLU()
  (12): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
  (13): Flatten(start_dim=1, end_dim=-1)
  (14): Linear(in_features=675, out_features=512, bias=True)
  (15): Dropout(p=0.3, inplace=False)
  (16): ReLU()
  (17): Linear(in_features

In [13]:
####################
#完善代码
####################

In [14]:
def validate():
    loss = 0
    accuracy = 0

####################
#完善代码
####################

In [17]:
def train():
    loss = 0
    accuracy = 0

####################
#完善代码
####################

In [18]:
####################
#完善代码   这里要完成20个epoch
####################

for epoch in range(epochs):
    print('Epoch: {}'.format(epoch))
####################
#完善代码
####################

Epoch: 0
Train - Loss: 169.8087 Accuracy: 0.9841
Valid - Loss: 178.3278 Accuracy: 0.9315
Epoch: 1
Train - Loss: 100.2671 Accuracy: 0.9912
Valid - Loss: 98.2832 Accuracy: 0.9653
Epoch: 2
Train - Loss: 58.6416 Accuracy: 0.9946
Valid - Loss: 99.5804 Accuracy: 0.9664
Epoch: 3
Train - Loss: 69.2122 Accuracy: 0.9942
Valid - Loss: 249.3665 Accuracy: 0.9373
Epoch: 4
Train - Loss: 49.6674 Accuracy: 0.9956
Valid - Loss: 138.5586 Accuracy: 0.9704
